scheduler/src/ratelimit.clj

```diff
@@ -0,0 +1,114 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns ratelimit
+  (:require [ratelimit.token-bucket-filter]
+            [clj-time.core]
+            [clj-time.coerce])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (earn-tokens! [this key]
+    "Account for any earned tokens for the requested key.")
+  (in-debt!? [this key]
+    "Boolean indicating if the rate limit is currently in debt. Earns tokens first, then checks to see if in debt.")
+  (spend-tokens! [this key tokens]
+    "Request this number of tokens. Does not block (or otherwise indicate we're in debt")
+  (time-until!? [this key]
+    "Time, in milliseconds, until this rate limiter is out of debt for this key. Earns tokens first.")
+  (is-enforcing? [this]
+    "Are we enforcing this policy?"))
+
+(defn now
+  []
+  (clj-time.coerce/to-long (clj-time.core/now)))
+
+(defn get-tbf
+  "Get or create a TBF for the requested key. Buckets start off full, so users whose first request is a lot of
+  fast requests don't get throttled immediately. This doesn't let users get much above their quota ---
+  The fact that we don't have a TBF says that we are either just starting up (not likely) or haven't
+  gotten anything from them in a long time. (Assuming the TTL is set at least as high as
+  (:bucket-size/:rate-per-minute)."
+  [limiter key]
+  (.get (:cache limiter) key (proxy [Callable] []
```

Comment:
It would be more readable to me if the last argument to `.get` was extracted into a local variable or we used `->>`